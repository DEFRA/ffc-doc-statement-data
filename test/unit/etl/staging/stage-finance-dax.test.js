const { v4: uuidv4 } = require('uuid')
const storage = require('../../../../app/storage')
const { stageFinanceDAX } = require('../../../../app/etl/staging/stage-finance-dax')
const { Readable } = require('stream')

jest.mock('uuid')
jest.mock('../../../../app/storage')
jest.mock('../../../../app/config/etl')
jest.mock('../../../../app/constants/tables')
jest.mock('../../../../app/etl/run-etl-process')

describe('stageFinanceDAX', () => {
  let runEtlProcess

  beforeAll(() => {
    runEtlProcess = require('../../../../app/etl/run-etl-process').runEtlProcess
    jest.spyOn(require('../../../../app/etl/run-etl-process'), 'runEtlProcess').mockResolvedValue()
  })

  test('should download the file and run the ETL process', async () => {
    const mockFile = 'Finance_Dax_SFI23/export.csv'
    const mockUuid = 'mock-uuid'

    uuidv4.mockReturnValue(mockUuid)
    const mockStreamData = 'CHANGE_TYPE,CHANGE_TIME,PKID,DT_INSERT\nINSERT,2021-01-01,1,2021-01-01\nUPDATE,2021-01-02,2,2021-01-02\n'
    const mockReadableStream = Readable.from(mockStreamData.split('\n'))
    storage.downloadFileAsStream = jest.fn().mockResolvedValue(mockReadableStream)

    await stageFinanceDAX()

    expect(storage.downloadFileAsStream).toHaveBeenCalledWith(mockFile)
    expect(runEtlProcess).toMatchSnapshot()
  })
})
