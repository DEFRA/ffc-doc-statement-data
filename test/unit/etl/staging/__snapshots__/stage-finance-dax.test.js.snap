// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stageFinanceDAX should download the file and run the ETL process 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "columns": [
          "CHANGE_TYPE",
          "CHANGE_TIME",
          "TRANSDATE",
          "ACCOUNTTYPE",
          "INVOICEID",
          "INVOICEDATE",
          "SCHEME",
          "FUND",
          "MARKETINGYEAR",
          "ACCOUNTNUM",
          "DELIVERYBODY",
          "ACCRUALSTATUS",
          "SETTLEMENTVOUCHER",
          "LINEAMOUNTEUR",
          "LINEAMOUNTMSTGBP",
          "TRANSACTIONTYPE",
          "JOURNALNUM",
          "JOURNALNAME",
          "ADMINDEBT",
          "VOUCHER",
          "MONTH",
          "AMOUNTEUR",
          "AMOUNTMSTGBP",
          "QUARTER",
          "IDVOUCHER",
          "TRANSTXT",
          "IRREGULARITYDEBT",
          "SECTIONTYPE",
          "TRANSRECID",
          "LJTRANSRECID",
          "GLACCOUNTENTRYRECID",
          "SETTLEMENTRECID",
          "RUNNUMBER",
          "REPORTINGTYPE",
          "CUSTVENDAC",
          "MODIFIEDDATETIME",
          "MODIFIEDBY",
          "CREATEDDATETIME",
          "CREATEDBY",
          "DATAAREAID",
          "PARTITION_FLD",
          "RECID",
          "EURUNNUMBER",
          "ACCRUALLEDGERDIMENSIONACCOUNT",
          "ACCRUALGJACCOUNTENTRYRECID",
          "CURRENCYCODE",
          "LEGACYFARMERACCOUNT",
          "CLAIMSETTLEMENTDATE",
          "CLAIMREFNUM",
          "ISEURELEVANT",
          "FUNDEDBY",
          "EUYEARSTARTDATE",
          "EUYEARENDDATE",
          "YEARENDRUNNUMBER",
          "REASON",
          "EXCHRATECALCTYPE",
          "POSTINGDATE",
          "DATAWAREHOUSERUNNUM",
          "CROSSCOMPLIANCE",
          "PENALTY",
          "EXCHFUNDEDEUR",
          "EXCHFUNDEDGBP",
          "FESTXNNUM",
          "FESTXNLINENUM",
          "ADVPAYFLAG",
          "DAXINVOICENUM",
          "DWHTXNTYPE",
          "RETENTION",
          "DUMMYT104",
          "F106GBP",
          "SCHEMEEXCHANGERATE",
          "AGREEMENTREFERENCE",
          "OEDATE",
          "SETTLEMENTVOUCHER1",
          "ANNEXIITRANSTYPE",
          "BPALLOCATIONTYPE",
          "SCHEMETYPE",
          "APPLICATIONREFERENCE",
        ],
        "excludedFields": [
          "accounttype",
          "accrualgjaccountentryrecid",
          "accrualledgerdimensionaccount",
          "accrualstatus",
          "admindebt",
          "advpayflag",
          "amounteur",
          "annexiitranstype",
          "applicationreference",
          "bpallocationtype",
          "claimrefnum",
          "claimsettlementdate",
          "createdby",
          "createddatetime",
          "crosscompliance",
          "currencycode",
          "dataareaid",
          "datawarehouserunnum",
          "daxinvoicenum",
          "deliverybody",
          "dummyt104",
          "dwhtxntype",
          "eurunnumber",
          "euyearenddate",
          "euyearstartdate",
          "exchfundedeur",
          "exchfundedgbp",
          "exchratecalctype",
          "f106gbp",
          "festxnnum",
          "festxnlinenum",
          "fundedby",
          "glaccountentryrecid",
          "idvoucher",
          "invoicedate",
          "irregularitydebt",
          "iseurelevant",
          "journalname",
          "journalnum",
          "legacyfarmeraccount",
          "lineamounteur",
          "ljtransrecid",
          "modifiedby",
          "modifieddatetime",
          "oedate",
          "partition",
          "penalty",
          "postingdate",
          "reason",
          "reportingtype",
          "retention",
          "runnumber",
          "schemeexchangerate",
          "sectiontype",
          "settlementrecid",
          "settlementvoucher1",
          "transactiontype",
          "transrecid",
          "transtxt",
          "voucher",
          "yearendrunnumber",
        ],
        "file": "Finance_Dax/export.csv",
        "fileStream": Readable {
          "_destroy": [Function],
          "_events": {
            "close": undefined,
            "data": undefined,
            "end": undefined,
            "error": undefined,
            "readable": undefined,
          },
          "_maxListeners": undefined,
          "_read": [Function],
          "_readableState": ReadableState {
            "awaitDrainWriters": null,
            "buffer": [],
            "bufferIndex": 0,
            "highWaterMark": 1,
            "length": 0,
            "pipes": [],
            Symbol(kState): 1052941,
          },
          Symbol(shapeMode): true,
          Symbol(kCapture): false,
        },
        "mapping": [
          {
            "column": "CHANGE_TYPE",
            "targetColumn": "changeType",
            "targetType": "varchar",
          },
          {
            "column": "CHANGE_TIME",
            "format": "DD-MM-YYYY HH24:MI:SS",
            "targetColumn": "changeTime",
            "targetType": "date",
          },
          {
            "column": "TRANSDATE",
            "format": "DD-MM-YYYY HH24:MI:SS",
            "targetColumn": "transdate",
            "targetType": "date",
          },
          {
            "column": "ACCOUNTTYPE",
            "targetColumn": "accounttype",
            "targetType": "number",
          },
          {
            "column": "INVOICEID",
            "targetColumn": "invoiceid",
            "targetType": "varchar",
          },
          {
            "column": "INVOICEDATE",
            "format": "DD-MM-YYYY HH24:MI:SS",
            "targetColumn": "invoicedate",
            "targetType": "date",
          },
          {
            "column": "SCHEME",
            "targetColumn": "scheme",
            "targetType": "varchar",
          },
          {
            "column": "FUND",
            "targetColumn": "fund",
            "targetType": "varchar",
          },
          {
            "column": "MARKETINGYEAR",
            "targetColumn": "marketingyear",
            "targetType": "varchar",
          },
          {
            "column": "ACCOUNTNUM",
            "targetColumn": "accountnum",
            "targetType": "varchar",
          },
          {
            "column": "DELIVERYBODY",
            "targetColumn": "deliverybody",
            "targetType": "varchar",
          },
          {
            "column": "ACCRUALSTATUS",
            "targetColumn": "accrualstatus",
            "targetType": "number",
          },
          {
            "column": "SETTLEMENTVOUCHER",
            "targetColumn": "settlementvoucher",
            "targetType": "varchar",
          },
          {
            "column": "LINEAMOUNTEUR",
            "targetColumn": "lineamounteur",
            "targetType": "number",
          },
          {
            "column": "LINEAMOUNTMSTGBP",
            "targetColumn": "lineamountmstgbp",
            "targetType": "number",
          },
          {
            "column": "TRANSACTIONTYPE",
            "targetColumn": "transactiontype",
            "targetType": "number",
          },
          {
            "column": "JOURNALNUM",
            "targetColumn": "journalnum",
            "targetType": "varchar",
          },
          {
            "column": "JOURNALNAME",
            "targetColumn": "journalname",
            "targetType": "varchar",
          },
          {
            "column": "ADMINDEBT",
            "targetColumn": "admindebt",
            "targetType": "number",
          },
          {
            "column": "VOUCHER",
            "targetColumn": "voucher",
            "targetType": "varchar",
          },
          {
            "column": "MONTH",
            "targetColumn": "month",
            "targetType": "varchar",
          },
          {
            "column": "AMOUNTEUR",
            "targetColumn": "amounteur",
            "targetType": "number",
          },
          {
            "column": "AMOUNTMSTGBP",
            "targetColumn": "amountmstgbp",
            "targetType": "number",
          },
          {
            "column": "QUARTER",
            "targetColumn": "quarter",
            "targetType": "varchar",
          },
          {
            "column": "IDVOUCHER",
            "targetColumn": "idvoucher",
            "targetType": "varchar",
          },
          {
            "column": "TRANSTXT",
            "targetColumn": "transtxt",
            "targetType": "varchar",
          },
          {
            "column": "IRREGULARITYDEBT",
            "targetColumn": "irregularitydebt",
            "targetType": "number",
          },
          {
            "column": "SECTIONTYPE",
            "targetColumn": "sectiontype",
            "targetType": "number",
          },
          {
            "column": "TRANSRECID",
            "targetColumn": "transrecid",
            "targetType": "number",
          },
          {
            "column": "LJTRANSRECID",
            "targetColumn": "ljtransrecid",
            "targetType": "number",
          },
          {
            "column": "GLACCOUNTENTRYRECID",
            "targetColumn": "glaccountentryrecid",
            "targetType": "number",
          },
          {
            "column": "SETTLEMENTRECID",
            "targetColumn": "settlementrecid",
            "targetType": "number",
          },
          {
            "column": "RUNNUMBER",
            "targetColumn": "runnumber",
            "targetType": "varchar",
          },
          {
            "column": "REPORTINGTYPE",
            "targetColumn": "reportingtype",
            "targetType": "number",
          },
          {
            "column": "CUSTVENDAC",
            "targetColumn": "custvendac",
            "targetType": "varchar",
          },
          {
            "column": "MODIFIEDDATETIME",
            "format": "DD-MM-YYYY HH24:MI:SS",
            "targetColumn": "modifieddatetime",
            "targetType": "date",
          },
          {
            "column": "MODIFIEDBY",
            "targetColumn": "modifiedby",
            "targetType": "varchar",
          },
          {
            "column": "CREATEDDATETIME",
            "format": "DD-MM-YYYY HH24:MI:SS",
            "targetColumn": "createddatetime",
            "targetType": "date",
          },
          {
            "column": "CREATEDBY",
            "targetColumn": "createdby",
            "targetType": "varchar",
          },
          {
            "column": "DATAAREAID",
            "targetColumn": "dataareaid",
            "targetType": "varchar",
          },
          {
            "column": "PARTITION_FLD",
            "targetColumn": "partition",
            "targetType": "varchar",
          },
          {
            "column": "RECID",
            "targetColumn": "recid",
            "targetType": "number",
          },
          {
            "column": "EURUNNUMBER",
            "targetColumn": "eurunnumber",
            "targetType": "varchar",
          },
          {
            "column": "ACCRUALLEDGERDIMENSIONACCOUNT",
            "targetColumn": "accrualledgerdimensionaccount",
            "targetType": "number",
          },
          {
            "column": "ACCRUALGJACCOUNTENTRYRECID",
            "targetColumn": "accrualgjaccountentryrecid",
            "targetType": "number",
          },
          {
            "column": "CURRENCYCODE",
            "targetColumn": "currencycode",
            "targetType": "varchar",
          },
          {
            "column": "LEGACYFARMERACCOUNT",
            "targetColumn": "legacyfarmeraccount",
            "targetType": "varchar",
          },
          {
            "column": "CLAIMSETTLEMENTDATE",
            "format": "DD-MM-YYYY HH24:MI:SS",
            "targetColumn": "claimsettlementdate",
            "targetType": "date",
          },
          {
            "column": "CLAIMREFNUM",
            "targetColumn": "claimrefnum",
            "targetType": "varchar",
          },
          {
            "column": "ISEURELEVANT",
            "targetColumn": "iseurelevant",
            "targetType": "number",
          },
          {
            "column": "FUNDEDBY",
            "targetColumn": "fundedby",
            "targetType": "number",
          },
          {
            "column": "EUYEARSTARTDATE",
            "format": "DD-MM-YYYY HH24:MI:SS",
            "targetColumn": "euyearstartdate",
            "targetType": "date",
          },
          {
            "column": "EUYEARENDDATE",
            "format": "DD-MM-YYYY HH24:MI:SS",
            "targetColumn": "euyearenddate",
            "targetType": "date",
          },
          {
            "column": "YEARENDRUNNUMBER",
            "targetColumn": "yearendrunnumber",
            "targetType": "varchar",
          },
          {
            "column": "REASON",
            "targetColumn": "reason",
            "targetType": "varchar",
          },
          {
            "column": "EXCHRATECALCTYPE",
            "targetColumn": "exchratecalctype",
            "targetType": "number",
          },
          {
            "column": "POSTINGDATE",
            "format": "DD-MM-YYYY HH24:MI:SS",
            "targetColumn": "postingdate",
            "targetType": "date",
          },
          {
            "column": "DATAWAREHOUSERUNNUM",
            "targetColumn": "datawarehouserunnum",
            "targetType": "varchar",
          },
          {
            "column": "CROSSCOMPLIANCE",
            "targetColumn": "crosscompliance",
            "targetType": "number",
          },
          {
            "column": "PENALTY",
            "targetColumn": "penalty",
            "targetType": "number",
          },
          {
            "column": "EXCHFUNDEDEUR",
            "targetColumn": "exchfundedeur",
            "targetType": "number",
          },
          {
            "column": "EXCHFUNDEDGBP",
            "targetColumn": "exchfundedgbp",
            "targetType": "number",
          },
          {
            "column": "FESTXNNUM",
            "targetColumn": "festxnnum",
            "targetType": "varchar",
          },
          {
            "column": "FESTXNLINENUM",
            "targetColumn": "festxnlinenum",
            "targetType": "number",
          },
          {
            "column": "ADVPAYFLAG",
            "targetColumn": "advpayflag",
            "targetType": "number",
          },
          {
            "column": "DAXINVOICENUM",
            "targetColumn": "daxinvoicenum",
            "targetType": "varchar",
          },
          {
            "column": "DWHTXNTYPE",
            "targetColumn": "dwhtxntype",
            "targetType": "varchar",
          },
          {
            "column": "RETENTION",
            "targetColumn": "retention",
            "targetType": "number",
          },
          {
            "column": "DUMMYT104",
            "targetColumn": "dummyt104",
            "targetType": "number",
          },
          {
            "column": "F106GBP",
            "targetColumn": "f106gbp",
            "targetType": "number",
          },
          {
            "column": "SCHEMEEXCHANGERATE",
            "targetColumn": "schemeexchangerate",
            "targetType": "number",
          },
          {
            "column": "AGREEMENTREFERENCE",
            "targetColumn": "agreementreference",
            "targetType": "varchar",
          },
          {
            "column": "OEDATE",
            "format": "DD-MM-YYYY HH24:MI:SS",
            "targetColumn": "oedate",
            "targetType": "date",
          },
          {
            "column": "SETTLEMENTVOUCHER1",
            "targetColumn": "settlementvoucher1",
            "targetType": "varchar",
          },
          {
            "column": "ANNEXIITRANSTYPE",
            "targetColumn": "annexiitranstype",
            "targetType": "number",
          },
          {
            "column": "BPALLOCATIONTYPE",
            "targetColumn": "bpallocationtype",
            "targetType": "number",
          },
          {
            "column": "SCHEMETYPE",
            "targetColumn": "schemetype",
            "targetType": "number",
          },
          {
            "column": "APPLICATIONREFERENCE",
            "targetColumn": "applicationreference",
            "targetType": "varchar",
          },
        ],
        "table": "etlStageFinanceDax",
        "transformer": [
          {
            "all": true,
            "column": "TRANSTXT",
            "find": "'.",
            "replace": "",
          },
        ],
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
